tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Analyzer.scala

```diff
@@ -350,6 +352,124 @@ final class Analyzer(semantics: Semantics, outputMode: OutputMode,
       loop(this)
     }
 
+    def tryLookupReflProxyMethod(proxyName: String): Option[MethodInfo] = {
+      if (!allowAddingSyntheticMethods) {
+        tryLookupMethod(proxyName)
+      } else {
+        /* The lookup for a target method in this code implements the
+         * algorithm defining `java.lang.Class.getMethod`. This mimics how
+         * reflective calls are implemented on the JVM, at link time.
+         *
+         * Caveat: protected methods are not ignored. This can only make an
+         * otherwise invalid reflective call suddenly able to call a protected
+         * method. It never breaks valid reflective calls. This could be fixed
+         * if the IR retained the information that a method is protected.
+         */
+
+        @tailrec
+        def loop(ancestorInfo: ClassInfo): Option[MethodInfo] = {
+          if (ancestorInfo ne null) {
+            ancestorInfo.methodInfos.get(proxyName) match {
+              case Some(m) =>
+                assert(m.isReflProxy && !m.isAbstract)
+                Some(m)
+
+              case _ =>
+                ancestorInfo.findProxyMatch(proxyName) match {
+                  case Some(target) =>
+                    val targetName = target.encodedName
+                    Some(ancestorInfo.createReflProxy(proxyName, targetName))
+
+                  case None =>
+                    loop(ancestorInfo.superClass)
+                }
+            }
```

Comment:
I admit I did not look at the previous implementation. I just looked anew at the output on the JVM after `cleanup`, saw it used `getMethod`, and looked at its specification.

With `Base` and `Sub` defined as above, we get this funny behavior:
```scala
scala> val x: { def foo(x: Option[Int]): Any } = new Sub
scala> x.foo(Some(1))
res1: Int = 1

scala> val y: { def foo(x: Option[String]): Any } = new Sub
scala> y.foo(Some("hello"))
res2: Any = 1
```
In both cases, `Sub.foo` is called. (There is no CCE because the `Option` param is not unwrapped.) I'm pretty sure this is consistent with my implementation. I'll add a test for it.